compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp

```diff
@@ -596,5 +597,35 @@ Optional<SmallVector<int64_t>> getWmmaNativeVectorSize(Operation *op) {
   return std::nullopt;
 }
 
+LogicalResult lowerAddressSpaceEnum(MLIRContext *ctx, ModuleOp &op,
+                                    unsigned globalAddrSpace,
+                                    unsigned workgroupAddrSpace,
+                                    unsigned privateAddrSpace) {
+  RewritePatternSet patterns(ctx);
+  TypeConverter typeConverter;
+  typeConverter.addConversion([](Type t) { return t; });
+  // NVVM uses alloca in the default address space to represent private
+  // memory allocations, so drop private annotations. NVVM uses address
+  // space 3 for shared memory. NVVM uses the default address space to
+  // represent global memory.
+  gpu::populateMemorySpaceAttributeTypeConversions(
+      typeConverter, [&](gpu::AddressSpace space) -> unsigned {
+        switch (space) {
+          case gpu::AddressSpace::Global:
+            return globalAddrSpace;
+          case gpu::AddressSpace::Workgroup:
+            return workgroupAddrSpace;
+          case gpu::AddressSpace::Private:
+            return privateAddrSpace;
+        }
+        llvm_unreachable("unknown address space enum value");
+        return 0;
+      });
+  gpu::populateMemorySpaceLoweringPatterns(typeConverter, patterns);
+  ConversionTarget target(*ctx);
+  gpu::populateLowerMemorySpaceOpLegality(target);
```

Comment:
Those patterns have been removed upstream, we need to do the memory space conversion during conversion to LLVM